### Importing Libraries


In [310]:
# Importing the python libraries
import pandas as pd
import numpy as ny
import seaborn as sns
from matplotlib import pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report, confusion_matrix, roc_auc_score, precision_recall_curve
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import GridSearchCV
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.model_selection import RandomizedSearchCV


In [311]:
# Load the Dataset.
customer_transaction_data = pd.read_csv(r"C:/Users/HP/Downloads/Fraudulent_Transaction_Detection_For_Finlora_Company/Finlora_Dataset/FinLora_Customer_Transaction_Dataset (1).csv")
customer_transaction_data.head()

,transaction_id,customer_id,timestamp,home_country,source_currency,dest_currency,channel,amount_src,amount_usd,fee,...,ip_risk_score,kyc_tier,account_age_days,device_trust_score,chargeback_history_count,risk_score_internal,txn_velocity_1h,txn_velocity_24h,corridor_risk,is_fraud
0,fee8542d-8ee6-4b0d-9671-c294dd08ed26,402cccc9-28de-45b3-9af7-cc5302aa1f93,2022-10-03 18:40:59.468549+00:00,US,USD,CAD,ATM,278.19,278.19,4.25,...,0.123,standard,263,0.522,0,0.223,0,0,0.0,0
1,bfdb9fc1-27fe-4a85-b043-4d813d679259,67c2c6b3-ef0a-4777-a3f1-c84a851bb6ad,2022-10-03 20:39:38.468549+00:00,CA,CAD,MXN,web,208.51,154.29,4.24,...,0.569,standard,947,0.475,0,0.268,0,1,0.0,0
2,fc855034-3ea5-4993-9afa-b511d93fe5e8,6d0d9b27-fa26-45f8-93b1-2df29d182d9c,2022-10-03 23:02:43.468549+00:00,US,USD,CNY,mobile,160.33,160.33,2.70,...,0.437,enhanced,367,0.939,0,0.176,0,0,0.0,0
3,2cf8c08e-42ec-444d-a755-34b9a2a0a4ca,7bd5200c-5d19-44f0-9afe-8b339a05366b,2022-10-04 01:08:53.468549+00:00,US,USD,EUR,mobile,59.41,59.41,2.22,...,0.594,standard,147,0.551,0,0.391,0,0,0.0,0
4,d907a74d-b426-438d-97eb-dbe911aca91c,70a93d26-8e3a-4179-900c-a4a7a74d08e5,2022-10-04 09:35:03.468549+00:00,US,USD,INR,mobile,200.96,200.96,3.61,...,0.121,enhanced,257,0.894,0,0.257,0,0,0.0,0


In [312]:
customer_transaction_data.tail()

,transaction_id,customer_id,timestamp,home_country,source_currency,dest_currency,channel,amount_src,amount_usd,fee,...,ip_risk_score,kyc_tier,account_age_days,device_trust_score,chargeback_history_count,risk_score_internal,txn_velocity_1h,txn_velocity_24h,corridor_risk,is_fraud
11395,b5ef3323-46d2-4f6c-9c19-43df4e55df48,8734e6a7-f34e-49f2-b559-1d62a756664c,2025-11-25 10:05:35.573611+00:00,US,USD,CAD,mobile,271.25,271.25,5.24,...,0.307,standard,742,0.653,0,0.302,1,2,0.00,0
11396,54f00f78-94aa-4a08-bf6b-12877aa47186,c18070a6-bdb9-4df3-9df7-5d92c0bff424,2025-11-26 07:09:56.573611+00:00,CA,CAD,NGN,web,537.17,397.51,8.86,...,0.891,low,9,0.173,2,0.730,3,6,0.00,1
11397,6a51f0e8-f5d1-4fe6-91a0-655fccd79fa5,2003e2a1-f3b1-4e2e-8d0e-488cfacf68ca,2025-11-27 06:19:11.573611+00:00,CA,CAD,INR,web,205.15,151.81,4.00,...,1.000,low,21,0.269,1,0.605,3,4,0.12,1
11398,4aad7389-2b62-4885-a23e-aa3ecd5cfaf9,8b1cf558-4ed7-48ee-b330-75db6efd4840,2025-11-28 00:53:28.573611+00:00,US,USD,PHP,mobile,78.03,78.03,1.96,...,0.298,enhanced,471,0.773,0,0.375,0,2,0.00,0
11399,fdffeb16-192a-4483-9b1e-9928e23269c2,b69010dc-ab0a-4fd2-a79e-65e4c6efbcd9,2025-11-29 20:10:47.573611+00:00,US,USD,NGN,web,1214.16,1214.16,19.34,...,1.000,low,58,0.187,0,0.646,4,6,0.25,1


In [313]:
customer_transaction_data.sample()

,transaction_id,customer_id,timestamp,home_country,source_currency,dest_currency,channel,amount_src,amount_usd,fee,...,ip_risk_score,kyc_tier,account_age_days,device_trust_score,chargeback_history_count,risk_score_internal,txn_velocity_1h,txn_velocity_24h,corridor_risk,is_fraud
3459,bf17979f-1e84-40cd-8cc2-9c37ef8c1bcf,67c2c6b3-ef0a-4777-a3f1-c84a851bb6ad,2023-10-23 10:07:06.468549+00:00,CA,CAD,MXN,mobile,65.47,48.45,1.71,...,0.628,standard,947,0.625,0,0.268,0,0,0.0,0


In [314]:
# Checking for number of rows and columns
customer_transaction_data.shape

(11400, 26)

In [315]:
# Checking for all the columns in the dataset
customer_transaction_data.columns


Index(['transaction_id', 'customer_id', 'timestamp', 'home_country',
       'source_currency', 'dest_currency', 'channel', 'amount_src',
       'amount_usd', 'fee', 'exchange_rate_src_to_dest', 'device_id',
       'new_device', 'ip_address', 'ip_country', 'location_mismatch',
       'ip_risk_score', 'kyc_tier', 'account_age_days', 'device_trust_score',
       'chargeback_history_count', 'risk_score_internal', 'txn_velocity_1h',
       'txn_velocity_24h', 'corridor_risk', 'is_fraud'],
      dtype='object')

In [316]:
# Classify Variable


identifier_variables = [
    'transaction_id',
    'customer_id',
    'device_id',
    'ip_address'
]

target_variable = [
    'is_fraud'
]

datetime_variables = [
    'timestamp'
]

categorical_variables = [
    'home_country',
    'source_currency',
    'dest_currency',
    'channel',
    'ip_country',
    'kyc_tier'
]

numerical_variables = [
    'amount_usd',
    'fee',
    'exchange_rate_src_to_dest',
    'ip_risk_score',
    'account_age_days',
    'device_trust_score',
    'chargeback_history_count',
    'risk_score_internal',
    'txn_velocity_1h',
    'txn_velocity_24h',
    'corridor_risk'
]

boolean_variables = [
    'new_device',
    'location_mismatch'
]

print("Identifier variables:")
print(identifier_variables)

print("\nTarget variable:")
print(target_variable)

print("\nDatetime variables:")
print(datetime_variables)

print("\nCategorical variables:")
print(categorical_variables)

print("\nNumerical variables:")
print(numerical_variables)

print("\nBoolean variables:")
print(boolean_variables)

Identifier variables:
['transaction_id', 'customer_id', 'device_id', 'ip_address']

Target variable:
['is_fraud']

Datetime variables:
['timestamp']

Categorical variables:
['home_country', 'source_currency', 'dest_currency', 'channel', 'ip_country', 'kyc_tier']

Numerical variables:
['amount_usd', 'fee', 'exchange_rate_src_to_dest', 'ip_risk_score', 'account_age_days', 'device_trust_score', 'chargeback_history_count', 'risk_score_internal', 'txn_velocity_1h', 'txn_velocity_24h', 'corridor_risk']

Boolean variables:
['new_device', 'location_mismatch']


In [317]:
# IDENTIFY PRIMARY AND FOREIGN KEYS


print("Checking transaction_id uniqueness...")

transaction_id_duplicates = (
    customer_transaction_data['transaction_id']
    .duplicated()
    .sum()
)

print("Duplicate transaction IDs:",
      transaction_id_duplicates)

print(
    "Unique transaction IDs:",
    customer_transaction_data['transaction_id'].nunique()
)

print(
    "Total records:",
    len(customer_transaction_data)
)

Checking transaction_id uniqueness...
Duplicate transaction IDs: 200
Unique transaction IDs: 11200
Total records: 11400


In [318]:
# Number of unique customers
print(
    "Unique customers:",
    customer_transaction_data['customer_id'].nunique()
)

# Number of transactions per customer
transactions_per_customer = (
    customer_transaction_data
    .groupby('customer_id')
    .size()
    .sort_values(ascending=False)
)

display(transactions_per_customer.head(10))

Unique customers: 1315


customer_id
402cccc9-28de-45b3-9af7-cc5302aa1f93    1510
d71c91b4-fee8-4104-9856-a5c6109a62e3    1355
7041b9c1-3719-4ca8-9a6b-811b47cea6c0    1345
6d0d9b27-fa26-45f8-93b1-2df29d182d9c    1066
af8ca4c4-8703-4c55-b66c-2b76cd70040d     915
7bd5200c-5d19-44f0-9afe-8b339a05366b     811
70a93d26-8e3a-4179-900c-a4a7a74d08e5     677
f7531a78-8bbe-4a85-b1e8-a0650edddc72     394
67c2c6b3-ef0a-4777-a3f1-c84a851bb6ad     360
23d9943d-57b1-42eb-9ef2-05eed7fd1957     347
dtype: int64

##### Observations:
###### Transaction_id is the strongest candidate for the primary key
###### customer_id → customer identifier / logical foreign key
###### device_id → device identifier
###### ip_address → network identifier

Because there is no separate customer table in this CSV, customer_id functions as a relational key within the integrated dataset rather than an enforced database foreign key.

In [319]:
# Checking for missing values
customer_transaction_data.isnull().sum()

transaction_id                 0
customer_id                    0
timestamp                     29
home_country                   0
source_currency                0
dest_currency                  0
channel                        0
amount_src                     0
amount_usd                   305
fee                          295
exchange_rate_src_to_dest      0
device_id                      0
new_device                     0
ip_address                   305
ip_country                   301
location_mismatch              0
ip_risk_score                  0
kyc_tier                     300
account_age_days               0
device_trust_score           295
chargeback_history_count       0
risk_score_internal            0
txn_velocity_1h                0
txn_velocity_24h               0
corridor_risk                  0
is_fraud                       0
dtype: int64

##### Checking for missing values in each column:
###### timestamp - 29 missing values
###### amount_usd - 305 missing values
###### fee - 295 missing values
###### ip_address - 305 missing values
###### ip_country - 301 missing values
###### kyc_tier - 300 missing values
###### device_trust_score - 295 missing values


In [320]:
customer_transaction_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11400 entries, 0 to 11399
Data columns (total 26 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   transaction_id             11400 non-null  object 
 1   customer_id                11400 non-null  object 
 2   timestamp                  11371 non-null  object 
 3   home_country               11400 non-null  object 
 4   source_currency            11400 non-null  object 
 5   dest_currency              11400 non-null  object 
 6   channel                    11400 non-null  object 
 7   amount_src                 11400 non-null  object 
 8   amount_usd                 11095 non-null  float64
 9   fee                        11105 non-null  float64
 10  exchange_rate_src_to_dest  11400 non-null  float64
 11  device_id                  11400 non-null  object 
 12  new_device                 11400 non-null  bool   
 13  ip_address                 11095 non-null  obj

In [321]:
customer_transaction_data.dtypes

transaction_id                object
customer_id                   object
timestamp                     object
home_country                  object
source_currency               object
dest_currency                 object
channel                       object
amount_src                    object
amount_usd                   float64
fee                          float64
exchange_rate_src_to_dest    float64
device_id                     object
new_device                      bool
ip_address                    object
ip_country                    object
location_mismatch               bool
ip_risk_score                float64
kyc_tier                      object
account_age_days               int64
device_trust_score           float64
chargeback_history_count       int64
risk_score_internal          float64
txn_velocity_1h                int64
txn_velocity_24h               int64
corridor_risk                float64
is_fraud                       int64
dtype: object

##### SUMMARY / OBSERVATIONS

###### The dataset contains 11,400 records and 26 variables, providing a substantial volume of transaction-level information for fraud detection analysis.
###### The field names generally follow standard variable-naming conventions, which supports consistency and readability throughout the data analysis and modelling process.
###### An initial assessment of the dataset identified data-type inconsistencies in some variables. In particular, the timestamp field is currently stored as an object/string data type, rather than as a datetime variable. This should be converted to an appropriate datetime format to support accurate temporal analysis and enable the extraction of useful features such as transaction hour, day, month, and day of the week.
###### The amount_src variable is also currently stored as an object data type, although it represents a transaction amount and should ideally be converted to a numeric format such as float. This conversion is necessary to support mathematical calculations, statistical analysis, outlier detection, and machine-learning feature engineering.
###### The missing-value assessment identified incomplete records across several variables. Specifically:
###### timestamp contains 29 missing values.
###### amount_usd contains 305 missing values.
###### ip_address contains 305 missing values.
###### fee contains 295 missing values.
###### device_trust_score contains 295 missing values.
###### ip_country contains 301 missing values.
###### kyc_tier contains 300 missing values.
###### These missing observations will need to be investigated and appropriately handled during the data-cleaning and preprocessing stage. The treatment method will depend on the nature and importance of each variable and may include imputation, use of an "Unknown" category, or removal of records where appropriate.
###### The target variable, is_fraud, is highly imbalanced. Approximately 91% of transactions are classified as legitimate, while only 9% are identified as fraudulent. This represents a significant class imbalance that could cause machine-learning algorithms to favour the majority class and consequently miss fraudulent transactions.
###### Given the class imbalance, accuracy should not be used as the primary model evaluation metric. A model could achieve a high accuracy score simply by predicting most transactions as legitimate while performing poorly at identifying actual fraud. Instead, the project should place greater emphasis on precision, recall, F1-score, Precision-Recall AUC (PR-AUC), and the confusion matrix, with particular attention to fraud recall.
###### Appropriate class-imbalance techniques should be investigated during model development. These may include class weighting, SMOTE/resampling techniques, and classification-threshold optimization. The final approach should balance the business cost of missed fraudulent transactions against the operational cost of excessive false-positive alerts

In [322]:
# Checking for label imbalance
customer_transaction_data['is_fraud'].value_counts(normalize=True).reset_index()

,is_fraud,proportion
0,0,0.912544
1,1,0.087456


In [323]:
# Potential Useful Fraud Variables

fraud_related_variables = [
    'amount_usd',
    'fee',
    'exchange_rate_src_to_dest',
    'new_device',
    'location_mismatch',
    'ip_risk_score',
    'kyc_tier',
    'account_age_days',
    'device_trust_score',
    'chargeback_history_count',
    'risk_score_internal',
    'txn_velocity_1h',
    'txn_velocity_24h',
    'corridor_risk',
    'channel',
    'home_country',
    'ip_country',
    'source_currency',
    'dest_currency',
    'timestamp'
]

print("Potentially useful fraud-related variables:")

for variable in fraud_related_variables:
    print("-", variable)

Potentially useful fraud-related variables:
- amount_usd
- fee
- exchange_rate_src_to_dest
- new_device
- location_mismatch
- ip_risk_score
- kyc_tier
- account_age_days
- device_trust_score
- chargeback_history_count
- risk_score_internal
- txn_velocity_1h
- txn_velocity_24h
- corridor_risk
- channel
- home_country
- ip_country
- source_currency
- dest_currency
- timestamp


In [324]:
# Variables Unsuitable for Modelling

potentially_unsuitable_variables = {
    'transaction_id': 'Unique transaction identifier; not a meaningful behavioural feature.',
    'customer_id': 'Customer identifier; high-cardinality identifier and potential overfitting risk.',
    'device_id': 'Device identifier; high-cardinality identifier and potential memorisation risk.',
    'ip_address': 'Highly granular identifier; direct modelling may cause high cardinality and privacy concerns.',
    'is_fraud': 'Target variable; must not be used as an input feature.'
}

for variable, reason in potentially_unsuitable_variables.items():
    print(f"{variable}: {reason}")

transaction_id: Unique transaction identifier; not a meaningful behavioural feature.
customer_id: Customer identifier; high-cardinality identifier and potential overfitting risk.
device_id: Device identifier; high-cardinality identifier and potential memorisation risk.
ip_address: Highly granular identifier; direct modelling may cause high cardinality and privacy concerns.
is_fraud: Target variable; must not be used as an input feature.


##### Observation:
is_fraud must never be included as a predictor because it is the target. Identifiers such as transaction ID, customer ID, device ID and raw IP address may also be excluded from the initial modelling feature set because they are identifiers rather than behavioural measurements and can introduce high cardinality, memorisation, leakage or privacy concerns.

##### Data Cleaning And Preparation


In [325]:
#converting data types of timestamp to datetime
customer_transaction_data['timestamp'] = pd.to_datetime(customer_transaction_data['timestamp'],errors='coerce') 
#converting data type of income_usd to float 
customer_transaction_data['amount_src'] = pd.to_numeric(customer_transaction_data['amount_src'],errors='coerce')

In [326]:
# Checking the data again to confirm the data type
customer_transaction_data.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11400 entries, 0 to 11399
Data columns (total 26 columns):
 #   Column                     Non-Null Count  Dtype              
---  ------                     --------------  -----              
 0   transaction_id             11400 non-null  object             
 1   customer_id                11400 non-null  object             
 2   timestamp                  11339 non-null  datetime64[ns, UTC]
 3   home_country               11400 non-null  object             
 4   source_currency            11400 non-null  object             
 5   dest_currency              11400 non-null  object             
 6   channel                    11400 non-null  object             
 7   amount_src                 11396 non-null  float64            
 8   amount_usd                 11095 non-null  float64            
 9   fee                        11105 non-null  float64            
 10  exchange_rate_src_to_dest  11400 non-null  float64            
 11  de

##### Data Type Conversion and Validation

###### During the data preprocessing stage, the timestamp and amount_src variables were converted from their original object/string formats into appropriate data types. The timestamp variable was converted to a datetime format using pd.to_datetime(), while amount_src was converted to a numeric format using pd.to_numeric(). Both conversions used the errors="coerce" argument to ensure that invalid or incorrectly formatted values were converted to missing values rather than causing the preprocessing process to fail.

###### Following the conversion, additional invalid records were identified. The number of missing values in timestamp increased from 29 to 61, indicating that 32 additional timestamp entries contained values that could not be correctly interpreted as dates or timestamps.

###### Similarly, amount_src initially contained no missing values; however, after conversion, 4 values were identified as invalid and converted to missing values. This suggests that these records contained improperly formatted, non-numeric, or corrupted transaction amounts.

###### These findings highlight the importance of data-type validation during preprocessing. The newly identified missing values were subsequently included in the data-cleaning process, where appropriate treatment methods were applied to maintain data quality and minimize the potential impact on subsequent exploratory analysis and machine-learning modelling.

In [327]:
# Rechecking for missing data
customer_transaction_data.isnull().sum()

transaction_id                 0
customer_id                    0
timestamp                     61
home_country                   0
source_currency                0
dest_currency                  0
channel                        0
amount_src                     4
amount_usd                   305
fee                          295
exchange_rate_src_to_dest      0
device_id                      0
new_device                     0
ip_address                   305
ip_country                   301
location_mismatch              0
ip_risk_score                  0
kyc_tier                     300
account_age_days               0
device_trust_score           295
chargeback_history_count       0
risk_score_internal            0
txn_velocity_1h                0
txn_velocity_24h               0
corridor_risk                  0
is_fraud                       0
dtype: int64


##### VARIABLES WITH MISSING VALUES THAT NEEDS IMPUTATION
###### During the data-cleaning stage, the variables containing missing values were reviewed and appropriate treatment methods were selected based on the nature and business meaning of each variable

######  amount usd: has 305 missing values. this will be filled by multiplying the amount_ src by the exchange rate of the currency. 
###### fee: has 295 missing values which will be filled by the median of channel variable if channel exists while other remaining fee missing values wil be filled using the overall median.
###### ip country: has 301 missing values which will be filled by the corresponding home_country 
###### kyc tier: has 300 missing values which will be filled with the most frequent kyc_tier(mode).
###### device trust score: has 295 missing values. This will be filled by the median of the group from grouping device_trust score by kyc_tier and new_ device variable or by the overal median of device.trust_score.
###### The timestamp variable which has 61 missing values and amount_src with 4 missing values will be dropped uSing the dropna method because the number of missing values in both variables is very small compared to the values we have on our dataset_ The ip_address variable has missing values of 305 cannot be filled with any value as each ip_address is a unique number so it will be dropped using the dropna method

In [328]:

#calculating exchange rates per currency .
exchange_rates = customer_transaction_data[customer_transaction_data['amount_usd'].notna()].groupby('source_currency').apply(

    lambda x: (x['amount_usd'] / x['amount_src']).mean() 
    ).to_dict() 
exchange_rates

C:\Users\HP\AppData\Local\Temp\ipykernel_14240\2185426304.py:2: FutureWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  exchange_rates = customer_transaction_data[customer_transaction_data['amount_usd'].notna()].groupby('source_currency').apply(


{'CAD': 0.7216095926871465,
 'GBP': 1.223441221648679,
 'USD': 0.9838730321259439}

In [329]:
# Filling the missing values of the amount_usd 
customer_transaction_data['amount_usd']=customer_transaction_data.apply(
    lambda row: row['amount_usd'] if pd.notna(row['amount_usd']) else row['amount_src'] * exchange_rates.get(row['source_currency'],1),
    axis=1
)


In [330]:
# Filling the missing values of the fee variables
if 'fee' in customer_transaction_data.columns:
   if 'channel' in customer_transaction_data.columns:
      customer_transaction_data['fee'] = customer_transaction_data.groupby('channel')['fee'].transform(lambda y: y.fillna(y.median()))
   customer_transaction_data['fee'] = customer_transaction_data['fee'].fillna(customer_transaction_data['fee'].median())

In [331]:
# Filling the missing values of ip_country variables
if {'ip_country', 'home_country'}.issubset(customer_transaction_data.columns):
    customer_transaction_data['ip_country'] = customer_transaction_data['ip_country'].fillna(customer_transaction_data['home_country'])

In [332]:
# Filling the missing values of device_trust_score variable
if 'device_trust_score' in customer_transaction_data.columns:
    if{'new_device', 'kyc_tier'}.issubset(customer_transaction_data.columns):
        customer_transaction_data['device_trust_score']*customer_transaction_data.groupby(['new_device', 'kyc_tier'])['device_trust_score'].transform(lambda x: x.fillna(x.median()))
        customer_transaction_data['device_trust_score']*customer_transaction_data['device_trust_score'].fillna(customer_transaction_data['device_trust_score'].median())


c:\Users\HP\Downloads\Fraudulent_Transaction_Detection_For_Finlora_Company\finlora_env\Lib\site-packages\numpy\lib\_nanfunctions_impl.py:1213: RuntimeWarning: Mean of empty slice
  return np.nanmean(a, axis, out=out, keepdims=keepdims)
c:\Users\HP\Downloads\Fraudulent_Transaction_Detection_For_Finlora_Company\finlora_env\Lib\site-packages\numpy\lib\_nanfunctions_impl.py:1213: RuntimeWarning: Mean of empty slice
  return np.nanmean(a, axis, out=out, keepdims=keepdims)


In [333]:
# Dropping the timestamp, amount_src and ip_address missing values
customer_transaction_data.dropna(subset=['timestamp', 'amount_src', 'ip_address'], inplace=True)

In [334]:
# Recheck for missing values
customer_transaction_data.isnull().sum()

transaction_id               0
customer_id                  0
timestamp                    0
home_country                 0
source_currency              0
dest_currency                0
channel                      0
amount_src                   0
amount_usd                   0
fee                          0
exchange_rate_src_to_dest    0
device_id                    0
new_device                   0
ip_address                   0
ip_country                   0
location_mismatch            0
ip_risk_score                0
kyc_tier                     0
account_age_days             0
device_trust_score           0
chargeback_history_count     0
risk_score_internal          0
txn_velocity_1h              0
txn_velocity_24h             0
corridor_risk                0
is_fraud                     0
dtype: int64

##### Steps for performing a data sanity check:
###### * ​Check for invalid numbers: Identify negative values in variables that should strictly be non-negative, such as age, monetary amounts, or transaction velocity.
###### * ​Check future timestamps: Ensure transaction timestamps do not contain dates or times set in the future relative to the dataset boundary or current time.
###### ​* Verify expected ranges: Confirm that numeric and categorical values fall within plausible, predefined logical ranges (e.g., percentages between 0–100, valid risk scores).
###### ​* Inspect variable consistency in the dataset: Cross-check related fields for logical conflicts (e.g., a transaction marked complete without a corresponding completed timestamp or valid account ID).

In [335]:
# Checking negative values in the numerical columns
negative_counts = {
    'amount_src': (customer_transaction_data['amount_src'] < 0).sum(),
    'amount_usd': (customer_transaction_data['amount_usd'] < 0).sum(),
    'fee': (customer_transaction_data['fee'] < 0).sum(),
    'device_trust_score': (customer_transaction_data['device_trust_score'] < 0).sum(),
    'txn_velocity_1h': (customer_transaction_data['txn_velocity_1h'] < 0).sum(),
    'txn_velocity_24h': (customer_transaction_data['txn_velocity_24h'] < 0).sum(),
    'risk_score_internal': (customer_transaction_data['risk_score_internal'] < 0).sum()    
}
negative_counts

{'amount_src': np.int64(100),
 'amount_usd': np.int64(0),
 'fee': np.int64(90),
 'device_trust_score': np.int64(190),
 'txn_velocity_1h': np.int64(190),
 'txn_velocity_24h': np.int64(0),
 'risk_score_internal': np.int64(0)}

In [336]:
# Drop the negative(less than < 0) variables
customer_transaction_data = customer_transaction_data.drop(customer_transaction_data[
    (customer_transaction_data['amount_src'] < 0) |
    (customer_transaction_data['fee'] < 0) |
    (customer_transaction_data['device_trust_score'] < 0) |
(customer_transaction_data['txn_velocity_1h'] < 0)] .index)

In [337]:
# Rechecking for negtive numbers
negative_counts = {
    'amount_src': (customer_transaction_data['amount_src'] < 0).sum(),
    'amount_usd': (customer_transaction_data['amount_usd'] < 0).sum(),
    'fee': (customer_transaction_data['fee'] < 0).sum(),
    'device_trust_score': (customer_transaction_data['device_trust_score'] < 0).sum(),
    'txn_velocity_1h': (customer_transaction_data['txn_velocity_1h'] < 0).sum(),
    'txn_velocity_24h': (customer_transaction_data['txn_velocity_24h'] < 0).sum(),
    'risk_score_internal': (customer_transaction_data['risk_score_internal'] < 0).sum()    
}
negative_counts

{'amount_src': np.int64(0),
 'amount_usd': np.int64(0),
 'fee': np.int64(0),
 'device_trust_score': np.int64(0),
 'txn_velocity_1h': np.int64(0),
 'txn_velocity_24h': np.int64(0),
 'risk_score_internal': np.int64(0)}

In [338]:
# Checking for futuristic timestamp
customer_transaction_data[customer_transaction_data['timestamp'] >pd.Timestamp.utcnow()]

,transaction_id,customer_id,timestamp,home_country,source_currency,dest_currency,channel,amount_src,amount_usd,fee,...,ip_risk_score,kyc_tier,account_age_days,device_trust_score,chargeback_history_count,risk_score_internal,txn_velocity_1h,txn_velocity_24h,corridor_risk,is_fraud


In [339]:
# Checking for location mismatch
customer_transaction_data['location_mismatch'].value_counts().reset_index()

,location_mismatch,count
0,False,9047
1,True,1793


In [340]:
# Reviewing the channel variable
customer_transaction_data['channel'].unique()

array(['ATM', 'web', 'mobile', 'WEB', ' web  ', 'MOBILE', 'unknown',
       'mobille', ' mobile  ', 'weeb', 'ATm', ' ATM  '], dtype=object)

In [341]:
# Formating the channel variable
customer_transaction_data['channel'] = customer_transaction_data['channel'].str.lower().str.strip()
customer_transaction_data['channel'].unique()

array(['atm', 'web', 'mobile', 'unknown', 'mobille', 'weeb'], dtype=object)

In [342]:
customer_transaction_data['channel'] = customer_transaction_data['channel'].replace({
    'weeb':'web',
    'mobille':'mobile',
})


In [357]:
customer_transaction_data['channel'] = customer_transaction_data['channel'].replace({'unknown':np.nan})
customer_transaction_data['channel'].unique()

array(['atm', 'web', 'mobile'], dtype=object)

In [344]:
# Reviewing the source currency variable
customer_transaction_data['source_currency'].unique()

array(['USD', 'CAD', 'GBP'], dtype=object)

In [345]:
# Reviewing the destination currency
customer_transaction_data['dest_currency'].unique()

array(['CAD', 'MXN', 'CNY', 'EUR', 'INR', 'GBP', 'PHP', 'NGN', 'USD'],
      dtype=object)

In [346]:
# Reviewing for kyc tier variabler
customer_transaction_data['kyc_tier'].unique() 

array(['standard', 'enhanced', 'low', ' standard  ', 'standrd',
       ' enhanced  ', 'STANDARD', 'unknown', 'enhancd', ' low  ',
       'ENHANCED', 'LOW'], dtype=object)

In [347]:
# Formating the kyc variable
customer_transaction_data['kyc_tier'] = customer_transaction_data['kyc_tier'].str.lower().str.strip()
customer_transaction_data['kyc_tier'].unique()

array(['standard', 'enhanced', 'low', 'standrd', 'unknown', 'enhancd'],
      dtype=object)

In [348]:
customer_transaction_data['kyc_tier'] = customer_transaction_data['kyc_tier'].replace({
    'standrd':'standard',
    'enhancd':'enhanced',
    'unknown':'np.nan'
})
customer_transaction_data['kyc_tier'].unique()

array(['standard', 'enhanced', 'low', 'np.nan'], dtype=object)

In [358]:
customer_transaction_data.isnull().sum()

transaction_id               0
customer_id                  0
timestamp                    0
home_country                 0
source_currency              0
dest_currency                0
channel                      0
amount_src                   0
amount_usd                   0
fee                          0
exchange_rate_src_to_dest    0
device_id                    0
new_device                   0
ip_address                   0
ip_country                   0
location_mismatch            0
ip_risk_score                0
kyc_tier                     0
account_age_days             0
device_trust_score           0
chargeback_history_count     0
risk_score_internal          0
txn_velocity_1h              0
txn_velocity_24h             0
corridor_risk                0
is_fraud                     0
dtype: int64

In [350]:
customer_transaction_data.dropna(inplace=True)
customer_transaction_data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10804 entries, 0 to 11399
Data columns (total 26 columns):
 #   Column                     Non-Null Count  Dtype              
---  ------                     --------------  -----              
 0   transaction_id             10804 non-null  object             
 1   customer_id                10804 non-null  object             
 2   timestamp                  10804 non-null  datetime64[ns, UTC]
 3   home_country               10804 non-null  object             
 4   source_currency            10804 non-null  object             
 5   dest_currency              10804 non-null  object             
 6   channel                    10804 non-null  object             
 7   amount_src                 10804 non-null  float64            
 8   amount_usd                 10804 non-null  float64            
 9   fee                        10804 non-null  float64            
 10  exchange_rate_src_to_dest  10804 non-null  float64            
 11  device_

In [351]:
# Reviewing the home country variable
customer_transaction_data['home_country'].unique()

array(['US', 'CA', 'UK', ' UK  ', ' US  ', 'unknown', ' CA  '],
      dtype=object)

In [352]:
# Formating the home country variable
customer_transaction_data['home_country'] = customer_transaction_data['home_country'].str.lower().str.strip()
customer_transaction_data['home_country'].unique()

array(['us', 'ca', 'uk', 'unknown'], dtype=object)

In [353]:
customer_transaction_data['home_country'] = customer_transaction_data['home_country'].replace({
        'unknown':'np.nan'
})
customer_transaction_data['home_country'].unique()

array(['us', 'ca', 'uk', 'np.nan'], dtype=object)

In [354]:
# Reviewing the ip country variable
customer_transaction_data['ip_country'].unique()

array(['US', 'CA', 'UK', ' US  ', 'unknown', ' CA  ', ' UK  '],
      dtype=object)

In [355]:
#  Formating the ip country variable
customer_transaction_data['ip_country'] = customer_transaction_data['ip_country'].str.lower().str.strip()
customer_transaction_data['ip_country'].unique()


array(['us', 'ca', 'uk', 'unknown'], dtype=object)

In [356]:
customer_transaction_data['ip_country'] = customer_transaction_data['ip_country'].replace({
    'unknown':'np.nan'
})
customer_transaction_data['ip_country'].unique()

array(['us', 'ca', 'uk', 'np.nan'], dtype=object)

##### FinLora Dataset Dictionary

| S/N | Column Name                 | Data Type   | Description                                                                                                     | Example               |
| --: | --------------------------- | ----------- | --------------------------------------------------------------------------------------------------------------- | --------------------- |
|   1 | `transaction_id`            | String      | Unique identifier assigned to each financial transaction.                                                       | `TXN10045892`         |
|   2 | `customer_id`               | String      | Unique identifier for the customer initiating or associated with the transaction.                               | `CUS002341`           |
|   3 | `timestamp`                 | Datetime    | Date and time when the transaction occurred.                                                                    | `2026-07-15 14:35:22` |
|   4 | `home_country`              | String      | Country where the customer account is registered.                                                               | `UK`                  |
|   5 | `source_currency`           | String      | Currency used by the sender.                                                                                    | `GBP`                 |
|   6 | `dest_currency`             | String      | Currency received by the recipient.                                                                             | `USD`                 |
|   7 | `channel`                   | Categorical | Digital channel through which the transaction was initiated.                                                    | `Mobile App`          |
|   8 | `amount_src`                | Float       | Transaction amount in the source currency.                                                                      | `500.00`              |
|   9 | `amount_usd`                | Float       | Transaction amount converted to USD.                                                                            | `620.50`              |
|  10 | `fee`                       | Float       | Transaction fee charged by the platform.                                                                        | `5.50`                |
|  11 | `exchange_rate_src_to_dest` | Float       | Exchange rate applied from the source currency to the destination currency.                                     | `1.24`                |
|  12 | `device_id`                 | String      | Unique identifier for the device used to initiate the transaction.                                              | `DEV98765`            |
|  13 | `new_device`                | Binary      | Indicates whether the device is new. `1 = Yes`, `0 = No`.                                                       | `1`                   |
|  14 | `ip_address`                | String      | IP address used for the transaction.                                                                            | `192.168.0.1`         |
|  15 | `ip_country`                | String      | Country detected from the transaction IP address.                                                               | `Germany`             |
|  16 | `location_mismatch`         | Binary      | Indicates whether the IP country differs from the customer's home country. `1 = Yes`, `0 = No`.                 | `1`                   |
|  17 | `ip_risk_score`             | Float       | Risk score assigned to the IP address.                                                                          | `0.72`                |
|  18 | `kyc_tier`                  | Integer     | Customer KYC verification level.                                                                                | `2`                   |
|  19 | `account_age_days`          | Integer     | Number of days since the customer's account was created.                                                        | `365`                 |
|  20 | `device_trust_score`        | Float       | Trust score assigned to the device used for the transaction.                                                    | `0.85`                |
|  21 | `chargeback_history_count`  | Integer     | Number of previous chargebacks associated with the customer or account.                                         | `1`                   |
|  22 | `risk_score_internal`       | Float       | Internal fraud-risk score assigned to the transaction or customer.                                              | `0.67`                |
|  23 | `txn_velocity_1h`           | Integer     | Number of transactions completed within the previous one-hour period.                                           | `3`                   |
|  24 | `txn_velocity_24h`          | Integer     | Number of transactions completed within the previous 24-hour period.                                            | `12`                  |
|  25 | `corridor_risk`             | Float       | Risk score associated with the transaction's country or currency corridor.                                      | `0.55`                |
|  26 | `is_fraud`                  | Binary      | Target variable indicating whether the transaction was identified as fraudulent. `1 = Fraud`, `0 = Legitimate`. | `0`                   |
